### PR TITLE
rkt driver plugin

### PR DESCRIPTION
**Pull request by preetapan**
Implements Rkt driver plugin, and ported over unit tests from client/driver/rkt_test.go

### CODE REVIEW

File: drivers/rkt/driver_test.go [hunk 1]
```diff
@@ -0,0 +1,750 @@
+// +build linux
+
+package rkt
+
```
Comment (preetapan) [inline drivers/rkt/driver_test.go]: TODOs here - 
fingerprint test 
stats test

File: plugins/drivers/driver.go [hunk 2]
```diff
@@ -111,6 +110,10 @@ type TaskConfig struct {
 	StderrPath      string
 }
 
+func (tc *TaskConfig) GetRawConfig() []byte {
```
Comment (nickethier) [inline plugins/drivers/driver.go]: Why is this needed?

File: drivers/rkt/driver.go [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+	capabilities = &drivers.Capabilities{
+		SendSignals: true,
+		Exec:        true,
+		FSIsolation: cstructs.FSIsolationChroot,
+	}
+
+	reRktVersion  = regexp.MustCompile(`rkt [vV]ersion[:]? (\d[.\d]+)`)
+	reAppcVersion = regexp.MustCompile(`appc [vV]ersion[:]? (\d[.\d]+)`)
+
+	_ drivers.DriverPlugin = (*RktDriver)(nil)
```
[the remark this answers is not in the document]
Comment (preetapan) [reply drivers/rkt/driver.go]: I put it here because it was useful during development as I implemented methods in the interface. Don't have a strong opinion, can move it to the test too

File: plugins/drivers/driver.go (hunk 2 above)
Comment (preetapan) [reply plugins/drivers/driver.go]: added when debugging msgpack unmarshalling, will remove

File: drivers/rkt/driver.go [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+	return capabilities, nil
+}
+
+func (r *RktDriver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go r.handleFingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (d *RktDriver) handleFingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
```
[the remark this answers is not in the document]
Comment (preetapan) [reply drivers/rkt/driver.go]: Done

File: drivers/rkt/driver.go [hunk 5]
```diff
@@ -0,0 +1,1007 @@
...
+// planned in the future
+type RktDriver struct {
+	// eventer is used to handle multiplexing of TaskEvents calls such that an
+	// event can be broadcast to all callers
+	eventer *eventer.Eventer
+
+	// config is the driver configuration set by the SetConfig RPC
+	config *Config
+
+	// tasks is the in memory datastore mapping taskIDs to execDriverHandles
```
Comment (schmichael) [inline drivers/rkt/driver.go]: update comment

File: drivers/rkt/driver.go [hunk 6]
```diff
@@ -0,0 +1,1007 @@
...
+	if trustPrefix != "" {
+		var outBuf, errBuf bytes.Buffer
+		cmd := exec.Command(rktCmd, "trust", "--skip-fingerprint-review=true", fmt.Sprintf("--prefix=%s", trustPrefix), fmt.Sprintf("--debug=%t", debug))
+		cmd.Stdout = &outBuf
+		cmd.Stderr = &errBuf
+		if err := cmd.Run(); err != nil {
+			return nil, nil, fmt.Errorf("Error running rkt trust: %s\n\nOutput: %s\n\nError: %s",
+				err, outBuf.String(), errBuf.String())
+		}
+		d.logger.Debug("added trust prefix", "trust_prefix", trustPrefix)
```
Comment (schmichael) [inline drivers/rkt/driver.go]: use the handle's logger or include `"task_name", cfg.Task.Name`

File: drivers/rkt/driver.go [hunk 7]
```diff
@@ -0,0 +1,1007 @@
...
+
+	// Setup port mapping and exposed ports
+	if len(cfg.Resources.NomadResources.Networks) == 0 {
+		d.logger.Debug("no network interfaces are available")
+		if len(driverConfig.PortMap) > 0 {
+			return nil, nil, fmt.Errorf("Trying to map ports but no network interface is available")
+		}
+	} else if network == "host" {
+		// Port mapping is skipped when host networking is used.
+		d.logger.Debug("driver.rkt: Ignoring port_map when using --net=host")
```
Comment (schmichael) [inline drivers/rkt/driver.go]:
```go
d.logger.Debug("ignoring port_map when using --net=host", "task_name", cfg.Task.Name)
```

File: drivers/rkt/driver.go [hunk 8]
```diff
@@ -0,0 +1,1007 @@
...
+			if mapped, ok := mergedPortMap[port.Label]; ok {
+				containerPort = mapped
+			} else {
+				// If the user doesn't have mapped a port using port_map, driver stops running container.
+				return nil, nil, fmt.Errorf("port_map is not set. When you defined port in the resources, you need to configure port_map.")
+			}
+
+			hostPortStr := strconv.Itoa(port.Value)
+
+			d.logger.Debug("driver.rkt: exposed port", "containerPort", containerPort)
```
Comment (schmichael) [inline drivers/rkt/driver.go]:
```go
d.logger.Debug("exposed port", "containerPort", containerPort, "task_name", cfg.Task.Name)
```

File: drivers/rkt/driver.go [hunk 9]
```diff
@@ -0,0 +1,1007 @@
...
+	rktDriverState := RktTaskState{
+		ReattachConfig: utils.ReattachConfigFromGoPlugin(pluginClient.ReattachConfig()),
+		Pid:            ps.Pid,
+		TaskConfig:     cfg,
+		StartedAt:      h.startedAt,
+		UUID:           uuid,
+	}
+
+	if err := handle.SetDriverState(&rktDriverState); err != nil {
+		d.logger.Error("failed to start task, error setting driver state", "error", err)
```
Comment (schmichael) [inline drivers/rkt/driver.go]:
```go
d.logger.Error("failed to start task, error setting driver state", "error", err, "task_name", cfg.Name)
```

File: drivers/rkt/driver.go [hunk 10]
```diff
@@ -0,0 +1,1007 @@
...
+	d.tasks.Set(cfg.ID, h)
+	go h.run()
+
+	// Do not attempt to retrieve driver network if one won't exist:
+	//  - "host" means the container itself has no networking metadata
+	//  - "none" means no network is configured
+	// https://coreos.com/rkt/docs/latest/networking/overview.html#no-loopback-only-networking
+	var driverNetwork *cstructs.DriverNetwork
+	if network != "host" && network != "none" {
+		d.logger.Debug("retrieving network information for pod", "pod", img, "UUID", uuid, "task", cfg.Name)
```
Comment (schmichael) [inline drivers/rkt/driver.go]: `"task" -> "task_name"`

File: drivers/rkt/driver.go [hunk 11]
```diff
@@ -0,0 +1,1007 @@
...
+		driverNetwork, err = rktGetDriverNetwork(uuid, mergedPortMap, d.logger)
+		if err != nil && !pluginClient.Exited() {
+			d.logger.Warn("network status retrieval for pod failed", "pod", img, "UUID", uuid, "task", cfg.Name, "error", err)
```
Comment (schmichael) [inline drivers/rkt/driver.go]: "task" -> "task_name"

(I kind of regret going with `task_name` because it's needlessly long, but we should keep it consistent.)

File: drivers/rkt/driver.go [hunk 12]
```diff
@@ -0,0 +1,1007 @@
...
+
+func (d *RktDriver) SignalTask(taskID string, signal string) error {
+	handle, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	sig := os.Interrupt
+	if s, ok := signals.SignalLookup[signal]; ok {
+		d.logger.Warn("signal to send to task unknown, using SIGINT", "signal", signal, "task_id", handle.taskConfig.ID)
```
Comment (schmichael) [inline drivers/rkt/driver.go]:
```go
d.logger.Warn("signal to send to task unknown, using SIGINT", "signal", signal, "task_name", handle.taskConfig.Name)
```

Although now you have me wondering if we should be logging the task ID everywhere. I think the answer is "no, it's an internal detail" but it's also the only way to differentiate restarts of a task....

File: drivers/rkt/driver.go [hunk 13]
```diff
@@ -0,0 +1,1007 @@
...
+		sig = s
+	}
+	return handle.exec.Signal(sig)
+}
+
+func (d *RktDriver) ExecTask(taskID string, cmdArgs []string, timeout time.Duration) (*drivers.ExecTaskResult, error) {
+	if len(cmdArgs) == 0 {
+		return nil, fmt.Errorf("error cmd must have atleast one value")
```
Comment (schmichael) [inline drivers/rkt/driver.go]: `atleast` -> `at least`

File: drivers/rkt/driver.go [hunk 14]
```diff
@@ -0,0 +1,1007 @@
...
+			ExitCode: ps.ExitCode,
+			Signal:   ps.Signal,
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-d.ctx.Done():
+		return
```
Comment (schmichael) [inline drivers/rkt/driver.go]: remove return

File: drivers/rkt/driver.go [hunk 15]
```diff
@@ -0,0 +1,1007 @@
...
+	} else {
+		result = &drivers.ExitResult{
+			ExitCode: ps.ExitCode,
+			Signal:   ps.Signal,
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return
```
Comment (schmichael) [inline drivers/rkt/driver.go]: remove return

File: drivers/rkt/driver.go [hunk 16]
```diff
@@ -0,0 +1,1007 @@
...
+	// a taskConfig within a job. It is returned in the TaskConfigSchema RPC
+	taskConfigSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"image":              hclspec.NewAttr("image", "string", true),
+		"command":            hclspec.NewAttr("command", "string", false),
+		"args":               hclspec.NewAttr("args", "list(string)", false),
+		"trust_prefix":       hclspec.NewAttr("trust_prefix", "string", false),
+		"dns_servers":        hclspec.NewAttr("dns_servers", "list(string)", false),
+		"dns_search_domains": hclspec.NewAttr("dns_search_domains", "list(string)", false),
+		"net":                hclspec.NewAttr("net", "list(string)", false),
+		"port_map":           hclspec.NewAttr("port_map", "list(map(string))", false),
```
[the remark this answers is not in the document]
Comment (preetapan) [reply drivers/rkt/driver.go]: I originally had it without the map but without an end to end test it was hard to prove that a pre nomad 0.9 rkt drive config block would parse correctly without modelling it as a slice.

File: drivers/rkt/driver.go [hunk 17]
```diff
@@ -0,0 +1,991 @@
...
+	// a taskConfig within a job. It is returned in the TaskConfigSchema RPC
+	taskConfigSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"image":              hclspec.NewAttr("image", "string", true),
+		"command":            hclspec.NewAttr("command", "string", false),
+		"args":               hclspec.NewAttr("args", "list(string)", false),
+		"trust_prefix":       hclspec.NewAttr("trust_prefix", "string", false),
+		"dns_servers":        hclspec.NewAttr("dns_servers", "list(string)", false),
+		"dns_search_domains": hclspec.NewAttr("dns_search_domains", "list(string)", false),
+		"net":                hclspec.NewAttr("net", "list(string)", false),
+		"port_map":           hclspec.NewAttr("port_map", "map(string)", false),
```
Comment (dadgar) [inline drivers/rkt/driver.go]: Should become: `"port_map":           hclspec.NewBlockAttrs("port_map", "string", false),` once this gets merged and r-clientv2 rebases: https://github.com/hashicorp/nomad/pull/4790
